clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java

```diff
@@ -1763,4 +1747,65 @@ private RecordAccumulator createTestRecordAccumulator(
             txnManager,
             new BufferPool(totalSize, batchSize, metrics, time, metricGrpName));
     }
+
+    private class SequentialRecordAccumulator extends RecordAccumulator {
```

Comment:
Could we move `AtomicInteger mockRandom` to class-level and then return `RecordAccumulator` with override  `createBuiltInPartitioner`? For example:
**RecordAccumulatorTest**
```java
    private AtomicInteger random = null;

    @AfterEach
    public void teardown() {
        this.metrics.close();
    }

    private RecordAccumulator createTestRecordAccumulator(
        TransactionManager txnManager,
        int deliveryTimeoutMs,
        int batchSize,
        long totalSize,
        Compression compression,
        int lingerMs
    ) {
        long retryBackoffMs = 100L;
        long retryBackoffMaxMs = 1000L;
        String metricGrpName = "producer-metrics";

        return new RecordAccumulator(
            logContext,
            batchSize,
            compression,
            lingerMs,
            retryBackoffMs,
            retryBackoffMaxMs,
            deliveryTimeoutMs,
            metrics,
            metricGrpName,
            time,
            new ApiVersions(),
            txnManager,
            new BufferPool(totalSize, batchSize, metrics, time, metricGrpName)) {
            @Override
            BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
                return new BuiltInPartitioner(logContext, topic, stickyBatchSize) {
                    @Override
                    int random() {
                        return random == null ? super.random() : random.getAndIncrement();
                    }
                };
            }
        };
    }
```

With that change, `testUniformBuiltInPartitioner` can define the `random` in the begining
```java
random = new AtomicInteger();
```

`testAdaptiveBuiltInPartitioner` can create `RecordAccumulator` with override `createBuiltInPartitioner` directly.
```java
        RecordAccumulator accum = new RecordAccumulator(logContext, batchSize, Compression.NONE, 0, 0L, 0L,
            3200, config, metrics, "producer-metrics", time, new ApiVersions(), null,
            new BufferPool(totalSize, batchSize, metrics, time, "producer-internal-metrics")) {
            @Override
            BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
                return new BuiltInPartitioner(logContext, topic, stickyBatchSize) {
                    @Override
                    int random() {
                        return random == null ? super.random() : random.getAndIncrement();
                    }
                };
            }
        };
```

Reply:
This is an excellent suggestion!, perfectly reproducing the "global modification" in the previous code.
I changed PR, plz review it @chia7712